### DISCUSSION TITLE
PDF report function augmentation

### DISCUSSION THREAD

**Discussion by JonasSchaub** [Steinbeck-Lab/ChemIcal-DatasEt-comparatoR#13] [Ideas]
Here, we can collect ideas and discuss how to augment the PDF export functionality to make it feel more like a report.

**Comment by hannbus** [comment]
Regarding the PDF export with the images on A4 format: As we discussed I tried to use BytesIO to change the matplotlib figure to png and give that to FPDF. Unfortunally, I could not make that work, because I get a PIL.PngImagePlugin.PngImageFile from BytesIO, which is not accepted by FPDF. Do you have any ideas how to get around the problem in FPDF?
As an alternative I also tried using reportlab for the PDF export instead of FPDF, but here I got some problems with the resolution of the figures (due to the need to resize the images they somehow lose their resolution).  If you don't have any idea for FPDF, maybe I can try working on this.

**Reply by OBrink** [reply]
Can you provide a reproducible code example for what you are doing with BytesIO and FPDF? The idea is that you write the image into the BytesIO object, and then you should be able to give the BytesIO object instead of a path to the FPDF methods.

**Reply by hannbus** [reply]
Here is what I tried with test_img as any mathplotlib.figure.Figure.

```
reader = io.BytesIO() 
test_img.savefig(reader) 
reader.seek(0) 
img = Image.open(reader) 

pdf = FPDF() 
pdf.add_page() 
pdf.image(img, 5, 5, 20)
```

It results in TypeError: unhashable type: 'PngImageFile'.

**Reply by OBrink** [reply]
According to FPDF's [documentation](https://pyfpdf.readthedocs.io/en/latest/reference/image/index.html), `pdf.image` is called with a path as the first argument. If I understand it right, you are trying to feed a PIL.Image object in there. Try replacing the last line with
`pdf.image(reader, 5, 5, 20)`
Then, you can delete `img = Image.open(reader) `. Do you really need `reader.seek(0) `?

**Reply by hannbus** [reply]
Yes, I also read that in the documentation and tried what you suggested but it leads to an AttributeError '_io.BytesIO' object has no attribute 'rfind'.
I don't think the reader.seek(0) should not be needed.

**Reply by OBrink** [reply]
I was able to reproduce the problem by running the following code block with the old version of[ FPDF which is not maintained anymore](https://github.com/reingart/pyfpdf). The solution is rather simple. Try using [FPDF2](https://github.com/PyFPDF/fpdf2). You can simply install it by running `pip install fpdf2`. You can use it the same way. When I run the code block below after installing FPDF2, everything works fine :)

```
import matplotlib.pyplot as plt
import io
from fpdf import FPDF

plt.plot([1, 2, 3, 4])
>>>>>> [<matplotlib.lines.Line2D object at 0x000001C3D132D520>]
plt.ylabel('some numbers')
>>>>>>Text(0, 0.5, 'some numbers')

pdf = FPDF()
pdf.add_page()

with io.BytesIO() as buffer:
    plt.savefig(buffer, format="png")
    pdf.image(buffer, 5, 5, 20, type='png')
```

**Reply by hannbus** [reply]
Thanks a lot, it is now working for me as well!

**Reply by JonasSchaub** [reply]
Great that you fixed that, thank you! :)
Just remember to test the deployment again if you update the project dependencies, is a point I wanted to at least mention 😅
